custom_components/rrd/__init__.py

```diff
@@ -123,47 +132,114 @@ def setup(hass, config):
-        ):  # all entities recently updated so lets store
+            # Get data value
+            sensor_state = hass.states.get(sensor_id)
             try:
-                ds_names, values = zip(
-                    *[(ds_name, value) for ds_name, _, value in entities.values()]
+                if sensor_state is None:
+                    _LOGGER.debug(
+                        "[%s] Skipping sensor %s, because value is unknown.", rrd_filename, sensor_id
+                    )
+                    raise Exception("Sensor has no value or not exists.")
+
+                sensor_value = sensor_state.state
+                # Convert value to integer, when type is COUNTER or DERIVE.
+                if (data_source[CONF_CF] in ["COUNTER", "DERIVE"]):
+                    sensor_value = round(float(sensor_value))
+            except:
+                _LOGGER.info(
+                    "[%s] sensor %s value will be stored as NaN.", rrd_filename, sensor_id
                 )
-                template = ":".join(ds_names)
-                timestamp = int(max(entities_last_changed))
-                values_string = ":".join([str(v) for v in values])
+                sensor_value = "NaN"
 
-                rrdtool.update(
-                    rrd_filename, f"-t{template}", f"{timestamp}:{values_string}"
-                )
-                _LOGGER.debug(
-                    "[%s] %s %s:%s", rrd_filename, template, timestamp, values_string,
-                )
-            except rrdtool.OperationalError as exc:
-                _LOGGER.error(exc)
-        else:
-            _LOGGER.debug("skipping <%s> until other DS update", ds_name)
+            # Add pais of name+value. Will be used as parameters for data save to rrd file.
+            ds_names.append(ds_name)
+            ds_values.append(str(sensor_value))
+
+        # Save to RRD file
+        try:
+            template = ":".join(ds_names)
+            timestamp = int(time.time())
+            values_string = ":".join(ds_values)
+
+            rrdtool.update(
+                rrd_filename, f"-t{template}", f"{timestamp}:{values_string}"
+            )
+            _LOGGER.debug(
+                "%s data added. ds=%s, values=%s:%s", rrd_filename, template, timestamp, values_string
+            )
+        except rrdtool.OperationalError as exc:
+            _LOGGER.error(exc)
+
+        # Schedule next update
+        schedule_next_update(database)
+
+
+    # Executed on Home assistant start
+    def start(_):
+        try:
+            for database in conf[CONF_DBS]:
+                # Run each database updating in own thread.
+                schedule_next_update(database)
+
+        except exc:
+            _LOGGER.error(exc)
+
+
+    # Stop updating all RRD files, because HASS is shutting down
+    def stop(_):
+        _LOGGER.debug("Stopping data updating")
+        # Stop to schedule any new update
+        hassIsShuttingDown = True
```

Comment:
I don't think we need `hassisshutttingDown`, all callback will be canceled before execution

Reply:
Good point. Thanks. Removed in https://github.com/dgomes/rrd_recorder/pull/3/commits/c78540dc5a5b0e9698ca9a0c31d707af34e76c75